Walkthrough of the Repository

quarpeeze opened the pull request:
This is a walkthrough that can help students navigate the repository in the beginning of the course. It should make the starting process easier and not only describe, but also make a quick demo of the functionality included in the pv211-utils.

1. For every project (Cranfield, ARQMath, CQADupStack, TREC) there is a brief description and demo code snippet. It lets the user see the data and extract it from the given dataset. 
2. There is a section on how to use datasets.py module to load data.
3. Text Preprocessing techniques such as lemmatization and stemming are demonstrated with examples.
4. Full demo of the systems package with code snippets to better understand how this package helps in retrieval tasks.
5. All the contents of the repo are (to the possible extent) mapped to Manning et al. 2008 book, with direct links to the book sections.
6. Extra links and useful sources (like articles about BERT) are included. There may be need to include more/filter out the current ones.

I tried to look at it from students' perspective and really make something I would rather have when I just started the course, so I hope it's at least a bit helpful :)

stefanik12 commented on walkthrough.ipynb:
```diff
@@ -0,0 +1,1950 @@
+    "Code implementations here are mostly linked to concepts from the book [Introduction to Information Retrieval](https://nlp.stanford.edu/IR-book/) by Manning, Raghavan, & Schütze (2008). You can download the full PDF version [HERE](https://nlp.stanford.edu/IR-book/pdf/irbookonlinereading.pdf).\n",
+    "\n",
+    "\n",
+    "- Starting off (cloning and setting up the repo) -> [Click here](#starting-off)\n",
```
I am not sure why, but the links do not work when looking at the notebook in Github, i.e. here: https://github.com/MIR-MU/pv211-utils/blob/1ad51a5874f5f8ce6d7c44a5ed7db0a422f72bca/walkthrough.ipynb

maybe there would be some easy fix, but I understand that the intended use is probably to download and access the notebook from some jupyter env, right

stefanik12 commented:
> I tried to look at it from students' perspective and really make something I would rather have when I just started the course, so I hope it's at least a bit helpful :)

I think it would be nice to add this -- or any disclosure of indended users at the beginning of the notebook. So that the reader knows (s)he's on the right place.

stefanik12 commented on walkthrough.ipynb:
```diff
@@ -0,0 +1,1950 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Cranfield Collection"
```
I was thinking maybe it would be nice to also walk the user over the **common concepts** of all the projects: i.e. that there are Documents, Judgements, IRSystem, etc., and how they interact together. Ideally also including the links to the base implementations in the repo.

stefanik12 replied:
It would also be nice to show how these objects are **used**: you already cover that at the end of the notebook:
```python
results_with_exp = list(retriever_with_exp.search(query))
for doc in results_with_exp:
    print(f"→ Doc {doc.document_id}: {doc.body[:80]}...")
```
...so maybe just copying a small demonstration of this kind to the beginning, before the introduction of specific projects (Cranfield, ...)?

stefanik12 commented on walkthrough.ipynb:
```diff
@@ -0,0 +1,1950 @@
+   "source": [
+    "# first let's install nltk if not already installed\n",
+    "\n",
+    "!pip install nltk"
```
Are we missing nltk from the pv211_utils dependencies?

quarpeeze commented:
> @quarpeeze, we'd love to have your contributions merged before the next semester so the students can already benefit from it. Do you plan to do any further changes?

Hi,
I apologize for not pushing any changes on time - after your last comments I started looking into why links in the notebook were not working and the other issues, but then got completely distracted.. Yes, I will for sure add my changes before Fri, 16th. I will also describe the changes here. In the meantime if there are any other ideas that you have on top (maybe something new came up after the Autumn semester) please let me know and I will work on it in the nearest days.

quarpeeze commented:
hey again, I covered all your comments and:

1. I read more info on these section links that don't work and tried it in several environments.. It didn't work anywhere at all outside my VScode,  so I think it's much better to just delete them and refer to sections without direct links

2. covering common concepts (implemented as classes) was a great idea and i was wrong to omit this in the first place! I added another section in the beginning. My idea was to show each concept through building a very simple IR system, using each concept one-by-one: building docs, defining query, adding judgements, using IRSystem etc..
 I think it's a nice thing to have - all of them wrapped in one micro pipeline in the very beginning of the walkthrough. what do you think?

3. small fixes: nltk in dependencies, adding some text to the beginning etc.

quarpeeze commented:
> @quarpeeze thank you for getting back to this! I've left two more comments, it would be great if you could take a look before we merge. Thanks!

Thank you for your comments! I looked into the errors in the notebook. Just letting you know that I'm still alive and I will push the updated version today (monday). Thank your for your time!

quarpeeze commented:
@stefanik12  sorry for the delay.

regarding the ValueError, it was only appearing when using tf-idf system, which uses multiprocessing with 'fork' method, which doesn't work on Windows (that's why it didn't work from my machine), so I tried doing something to it without changing much, but couldn't find a simple and reliable solution to run it on every OS without a problem (i lack knowledge to do it properly)

from my own research: there are potential solutions to that (e.g. use threading instead or somehow add spawn method (tried it but it crashed) or get rid of multiprocessing...) But if we just leave it as it is, it has to work fine on Mac/Linux as well as in google colab environment (tried that myself), so fixing it for Windows specially only for single TF-IDF system might not worth that, don't you think?

SO what I did:

1) I now pushed the notebook after rerunning TF-iDF cells in Colab env, all outputs are there, and there are NO ERRORS in the outputs.
2) in the beginning I added a cell to install the repo when running in Colab
3) In the TF-iDF section I added a note to warn Windows users of the possible error related to multuprocessing and the above problem.
4) I removed the other mentions of GitLab there.

Please let me know if there are any issues and I will get to it immediately.
